app/controllers/composer_controller.rb

```diff
@@ -0,0 +1,166 @@
+      .index_by(&:username_lower)
+
+    groups = Group
+      .visible_groups(current_user)
+      .where(name: names)
+      .index_by(&:name)
+
+    mentionable_group_ids = Group
+      .mentionable(current_user, include_public: false)
+      .where(name: names)
+      .pluck(:id)
+      .to_set
+
+    members_visible_group_ids = Group
+      .members_visible_groups(current_user)
+      .where(name: names)
+      .pluck(:id)
+      .to_set
+
+    topic_muted_by = if topic.present?
+      TopicUser
+        .where(topic: topic)
+        .where(user_id: users.values.map(&:id))
+        .where(notification_level: TopicUser.notification_levels[:muted])
+        .pluck(:user_id)
+        .to_set
+    else
+      Set.new
+    end
+
+    topic_allowed_user_ids = if allowed_names.present?
+      User
+        .where(username_lower: allowed_names.map(&:downcase))
+        .pluck(:id)
+        .to_set
+    elsif topic&.private_message?
+      TopicAllowedUser
+        .where(topic: topic)
+        .pluck(:user_id)
+        .to_set
+    end
+
+    topic_allowed_group_ids = if allowed_names.present?
+      Group
+        .messageable(current_user)
+        .where(name: allowed_names)
+        .pluck(:id)
+        .to_set
+    elsif topic&.private_message?
+      TopicAllowedGroup
+        .where(topic: topic)
+        .pluck(:group_id)
+        .to_set
+    end
+
+    names.each do |username|
+      user = users[username]
+      next if user.blank?
+
+      reason = if topic && !user.guardian.can_see?(topic)
+        topic.private_message? ? :private : :category
+      elsif allowed_names.present? && !is_user_allowed?(user, topic_allowed_user_ids, topic_allowed_group_ids)
+        # This would normally be handled by the previous if, but that does not work for new private messages.
+        :private
+      elsif topic_muted_by.include?(user.id)
+        :muted_topic
+      elsif topic&.private_message? && !is_user_allowed?(user, topic_allowed_user_ids, topic_allowed_group_ids)
+        # Admins can see the topic, but they will not be mentioned if they were not invited.
+        :not_allowed
+      end
+
+      # Regular users can see only basic information why the users cannot see the topic.
+      reason = nil if !guardian.is_staff? && reason != :private && reason != :category
+
+      user_reasons[username] = reason if reason.present?
+    end
+
+    names.each do |name|
+      group = groups[name]
+      next if group.blank?
+
+      group_reasons[name] = if !mentionable_group_ids.include?(group.id)
+        :not_mentionable
+      elsif (topic&.private_message? || allowed_names.present?) && !topic_allowed_group_ids.include?(group.id)
+        :not_allowed
+      end
+    end
+
+    if topic && names.include?(SiteSetting.here_mention) && guardian.can_mention_here?
+      here_count = PostAlerter.new.expand_here_mention(topic.first_post, exclude_ids: [current_user.id]).size
+    end
+
+    serialized_groups = groups.values.map do |group|
```

Comment:
```suggestion
    serialized_groups = groups.values.reduce({}) do |hash, group|
```

Minor but you can just reduce this to a hash straight away without having to return an array of arrays and then converting it to a hash.